### ISSUE TITLE
[$250 - Payment Due 9/28] IOS - Payments - The button "Add payment method" is not visible when scrolled up 

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#8929]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open app
2. Log in any account
3. Setting< Payments
4. Scroll up the app

## Expected Result:
The button "Add payment method" is visible when scrolled up 

## Actual Result:
The button "Add payment method" is not visible when scrolled up 

## Workaround:
Uknown
## Platform:

Where is this issue occurring?

- iOS

**Version Number:** 1.1.57.12

**Reproducible in staging?:** Yes

**Reproducible in production?:** No (new feature)

**Email or phone of affected tester (no customers):** any

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/167727656-2101dcc6-2690-4826-84dc-a7c8ab125312.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by laurenreidexpensify** [comment]
https://www.upwork.com/jobs/~01d4896d71392866b8

**Comment by parasharrajat** [comment]
Does not seem like an real issue. As soon as the user leaves the scroll, button moves to the bottom.

**Comment by JosueEchandiaAsto** [comment]
@parasharrajat 
### Proposal

### Issue
Extra space displayed when used non-quoted and quoted than non-quoted text. (Android and iOS)

### Fix
Scrollable elements in iOS have a bounce property, setting this to false disables this behavior.
`bounces={false}` to FlatList in PaymentMethodList.

Commit: https://github.com/Expensify/App/compare/main...JosueEchandiaAsto:Fix-Issue-8929

**Comment by parasharrajat** [comment]
> Extra space displayed when used non-quoted and quoted than non-quoted text. (Android and iOS)

How is this related?

Also, I don't like the proposed solution l.

**Comment by mananjadhav** [comment]
@JosueEchandiaAsto I would agree with @parasharrajat. 

> button should just be outside the scrollview and anchored to the bottom of the page

And for uniformity with other screens, I would go ahead with @roryabraham's recommendation for the expected behaviour.

**Comment by jayeshmangwani** [comment]
#### Proposal:

we have to remove the Button from the FlatList and need to add addPaymentMethod button wrapped in the FixedFooter component outside of the FlatList, 

https://github.com/Expensify/App/blob/d8e4adae05abbd45ca72fe5c80298686456e9e63/src/pages/settings/Payments/PaymentMethodList.js#L238-L242

```diff
+ import FixedFooter from '../../../components/FixedFooter';

+<>
                <FlatList
                    data={this.createPaymentMethodList()}
                    renderItem={this.renderItem}
                    keyExtractor={item => item.key}
                />
+                <FixedFooter>
+                   <Button
+                        text={this.props.translate('paymentMethodList.addPaymentMethod')}
+                        icon={Expensicons.CreditCard}
+                        onPress={this.props.onPress}
+                        isDisabled={this.props.isLoadingPayments}
-                        style={item.style} // we are wrapping the button FixedFooter so no need to add style here
+                        iconStyles={[styles.mr4]}
+                        success={true}
+                       shouldShowRightIcon={true}
+                      extraLarge
+                   />
+                </FixedFooter>
+            </>
```
    
we can remove these lines from here, no need to push data for FlatList
    https://github.com/Expensify/App/blob/d8e4adae05abbd45ca72fe5c80298686456e9e63/src/pages/settings/Payments/PaymentMethodList.js#L158-L169
    
we can remove these lines from here, no need to add this condition for renderItem


https://github.com/Expensify/App/blob/d8e4adae05abbd45ca72fe5c80298686456e9e63/src/pages/settings/Payments/PaymentMethodList.js#L211-L225
   
   
  

https://user-images.githubusercontent.com/35371050/173761293-36589f9a-5efc-4204-bac2-032b9e6bd715.mov

![Simulator Screen Shot - iPhone 8 - 2022-06-15 at 12 00 11](https://user-images.githubusercontent.com/35371050/173761320-ebe85a10-0153-4f4d-98a1-d40939ddf8b9.png)

**Comment by laurenreidexpensify** [comment]
I dunno why melvin thinks this is Daily

**Comment by parasharrajat** [comment]
@laurenreidexpensify As soon a new proposal is added, issues are moved to daily.

**Comment by mananjadhav** [comment]
@jayeshmangwani While your solution does fix the button at the bottom, it breaks the Add Payment button event. The payment options are not visible.

![image](https://user-images.githubusercontent.com/3069065/173935161-88f62b8b-9d21-47d6-85e3-90da35648333.png)

**Comment by JosueEchandiaAsto** [comment]
@mananjadhav @parasharrajat Is it true. I didn't see @roryabraham's comment, can I elaborate another proposal with this additional comment?

**Comment by JosueEchandiaAsto** [comment]
with @roryabraham's recommendation

### Proposal

### Issue
The button should just be outside the scrollview and anchored to the bottom of the page

### Fix
Commit: https://github.com/Expensify/App/commit/ee3bf75c4f693a56bb55c24356d81a2a82e8585f

First, we remove the following lines of code, so as not to have the 'Add payment method' button, additionally the conditional  `if (!this.props.shouldShowAddPaymentMethodButton)` is deleted because it is unnecessary.
https://github.com/Expensify/App/blob/594456eb6dd8db3f250d5dd5dea74f58780d82c9/src/pages/settings/Payments/PaymentMethodList.js#L154-L169

This method would look like this
https://github.com/Expensify/App/blob/ee3bf75c4f693a56bb55c24356d81a2a82e8585f/src/pages/settings/Payments/PaymentMethodList.js#L143-L156

Now in `render()` add the button to add payment methods and we validate `this.props.shouldShowAddPaymentMethodButton` so as not to affect the current logic.
https://github.com/Expensify/App/blob/594456eb6dd8db3f250d5dd5dea74f58780d82c9/src/pages/settings/Payments/PaymentMethodList.js#L236-L244

This method would look like this
https://github.com/Expensify/App/blob/ee3bf75c4f693a56bb55c24356d81a2a82e8585f/src/pages/settings/Payments/PaymentMethodList.js#L219-L245

So far, the expected behavior in iOS/Android is there, but some more adjustments are necessary for the popoper to work in web/desktop.

In Payment methods, the menu is shown below.
<img width="341" alt="" src="https://user-images.githubusercontent.com/16966456/174002842-761845ea-43c4-4b9d-9dbd-ac9e6b4ce6df.png">

In Add Payment Methods, the menu would be shown above.
<img width ="341" alt="" src="https://user-images.githubusercontent.com/16966456/174002833-18ce6094-2796-460e-91a4-e536c64e241f.png">

To do this, anchorPositionBottom and a new method setPositionAddPaymentMethod in BasePaymentsPage have been added.
https://github.com/Expensify/App/blob/ee3bf75c4f693a56bb55c24356d81a2a82e8585f/src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js#L48

https://github.com/Expensify/App/blob/ee3bf75c4f693a56bb55c24356d81a2a82e8585f/src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js#L121-L132

https://github.com/Expensify/App/blob/ee3bf75c4f693a56bb55c24356d81a2a82e8585f/src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js#L319-L336

https://github.com/Expensify/App/blob/ee3bf75c4f693a56bb55c24356d81a2a82e8585f/src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js#L403-L441

The existing method `setPositionAddPaymentMenu` is also modified
https://github.com/Expensify/App/blob/ee3bf75c4f693a56bb55c24356d81a2a82e8585f/src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js#L106-L120

Additionally, in the button component it is necessary to add `styles.pointerEventsNone` so that the child elements of the Button are not clickable and the location of the button is correctly calculated when consulting `getClickedElementLocation`.
https://github.com/Expensify/App/blob/ee3bf75c4f693a56bb55c24356d81a2a82e8585f/src/components/Button.js#L200-L205
https://github.com/Expensify/App/blob/594456eb6dd8db3f250d5dd5dea74f58780d82c9/src/libs/getClickedElementLocation/index.js#L7-L9

### Result
iOS

https://user-images.githubusercontent.com/16966456/174000615-f7e8dcee-1262-4f94-bf24-dc138c48adfd.mp4

Android

https://user-images.githubusercontent.com/16966456/174000459-e63fdd5e-b484-456f-8b93-0c7ea7d71e48.mov

Web

https://user-images.githubusercontent.com/16966456/174000797-1cc7db21-2574-404d-bba0-32905040c41f.mov

iOS/Web

https://user-images.githubusercontent.com/16966456/174000887-bda16ed0-3b33-4e89-9417-3440e504cb26.mp4

**Comment by mananjadhav** [comment]
@JosueEchandiaAsto Your updated proposal is in line with what is expected. Can you help with the following?

1. We're setting `bottom: this.state.anchorPostionBottom` to all the `Popovers`. Is it needed?
2. We're setting `anchorPositionBottom` to `null`. Any specific reason why?
3. We're updating only one `setPositionAddPaymentMenu` call to `setPositionAddPaymentMethod`, but not the others, why?
4. We're adding `styles.pointerEventsNone` to Button with icon, I am not sure why this change is required?

**Comment by JosueEchandiaAsto** [comment]
@mananjadhav 

> We are setting the background: this.state.anchorPostionBottom for all Popovers. Necessary?

You are right, it is not necessary
> We are setting AnchorPositionBottom to null. Any specific reason why?

I have changed it in the update.
> We are updating only one setPositionAddPaymentMenu call to setPositionAddPaymentMethod, but not the others, why?

`setPositionAddPaymentMenu` is for all the menus that appear below the buttons.
`setPositionAddPaymentMethod` is only for the add payment method button because it should appear above it.

> We are adding styles.pointerEventsNone to the button with icon, not sure why this change is required.

It was a way to prevent the menu from overlapping the button, in the update I have made this change so as not to touch the button component.

### Update of proposal

1. We change the position left to position right in all popovers, because when the browser is resized the menu doesn't position itself well.
2. Avoid using `styles.pointerEventsNone`3. 
3. anchorPositionBottom is only used for add payment method menu.4. 

Commit https://github.com/Expensify/App/commit/a1c89775404212d8fd1b7ebf551ba705737db971

Result:

https://user-images.githubusercontent.com/16966456/174669470-446be63d-ee89-4082-a970-0c74242f2e08.mov

**Comment by mananjadhav** [comment]
@JosueEchandiaAsto's proposal looks good. Some minor code quality can be fixed in the PR feedback cycle.

@roryabraham What are your thoughts?

🎀 👀 🎀
C+ reviewed

**Comment by roryabraham** [comment]
The proposal itself was a bit hard to follow, especially with the few edits, but it's good enough for me 👍

**Comment by JosueEchandiaAsto** [comment]
@mananjadhav
How to proceed? My proposal also solves this problem  https://github.com/Expensify/App/issues/9283 that PR already has. Do I wait for such a proposal to be implemented or do I remove my solution for the overlay?

**Comment by mananjadhav** [comment]
I feel you can ignore that part of the change on the PR. It'll mostly get merged by Monday-Tuesday. So shouldn't be a problem.

**Comment by roryabraham** [comment]
PR is in review

**Comment by laurenreidexpensify** [comment]
PR still in review

**Comment by roryabraham** [comment]
Thanks for your patience here everyone, this PR somehow slid through the cracks and I failed to review it promptly. It's got my attention now and we should be pretty close to a mergeable state with it.

**Comment by laurenreidexpensify** [comment]
@MitchExpensify  👋 am OOO til 12 August so reassigning. Thanks!

**Comment by MitchExpensify** [comment]
Enjoy @laurenreidexpensify !

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by MitchExpensify** [comment]
Do we need to move forward with [the PR](https://github.com/Expensify/App/pull/9622) if the issue not reproducible?

**Comment by roryabraham** [comment]
@MitchExpensify it might depend on what iOS device is used. I think we should move forward with the PR for the sake of code consistency and cleanup too. @JosueEchandiaAsto when can we expect an update?

**Comment by mananjadhav** [comment]
@JosueEchandiaAsto Please update on the PR. We've waited for more than 3 weeks now for an update now.

**Comment by mananjadhav** [comment]
@JosueEchandiaAsto We've sent multiple reminders on the PR and here. Please send us an update on the PR by Monday EOD? Otherwise we'll have to rescind the job.


cc - @roryabraham @MitchExpensify

**Comment by MitchExpensify** [comment]
Good call @mananjadhav. Lets now move to rescind the job without update from @JosueEchandiaAsto after the bump above. I'll double the price and post to Slack once we've rescinded it

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by MitchExpensify** [comment]
Scratch the doubling next step.. this hasn't been reproducible for 2 weeks - closing

**Comment by mananjadhav** [comment]
@MitchExpensify Afaik the issue is reproducible on smaller devices. Also the `Add Payment Button` is inconsistent with the rest of the UIs where we dock the button at the bottom.

**Comment by JosueEchandiaAsto** [comment]
Oh, I'm so sorry!. I was focused on other projects, and I think I accidentally disabled issue notifications.
If that's all right, I want to finish this build and send a PR update in tomorrow for review.

**Comment by MitchExpensify** [comment]
@JosueEchandiaAsto do you have an update on the PR?

**Comment by JosueEchandiaAsto** [comment]
@roryabraham Requested changes are done https://github.com/Expensify/App/pull/9622

**Comment by mananjadhav** [comment]
Thanks @JosueEchandiaAsto let me take a look at it today.

**Comment by JosueEchandiaAsto** [comment]
Again, sorry for the delay.  I will be pending so that this issue can be fixed as soon as possible

**Comment by mananjadhav** [comment]
An update here, PR is reviewed from C+, should be merged soon

**Comment by mananjadhav** [comment]
This was deployed to production 2 days back, but no automation comment on the merge here in the issue? And no update on payment date too.

**Comment by MitchExpensify** [comment]
Good catch @mananjadhav! I've made a note to check for regressions on Wednesday before issuing payment

**Comment by MitchExpensify** [comment]
Still saying "6 days" on [the merge to prod](https://github.com/Expensify/App/pull/9622#issuecomment-1255413916) so holding until it hits 7
